webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueCompilationCache.scala

```diff
@@ -0,0 +1,167 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * In accordance with the terms of section 7 (7. Additional Terms.) of
+ * the GNU General Public License version 3, the copyright holders add
+ * the following Additional permissions:
+ * Notwithstanding to the terms of section 5 (5. Conveying Modified Source
+ * Versions) and 6 (6. Conveying Non-Source Forms.) of the GNU General
+ * Public License version 3, when you create a Related Module, this
+ * Related Module is not considered as a part of the work and may be
+ * distributed under the license agreement of your choice.
+ * A "Related Module" means a set of sources files including their
+ * documentation that, without modification of the Source Code, enables
+ * supplementary functions or services in addition to those offered by
+ * the Software.
+ *
+ * Rudder is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+ *
+ *************************************************************************************
+ */
+
+package com.normation.rudder.ncf
+
+import com.normation.errors.IOResult
+import com.normation.inventory.domain.Version
+import com.normation.rudder.batch.UpdateCompilationStatus
+import com.normation.rudder.domain.logger.ConfigurationStatusLoggerPure
+import net.liftweb.common.SimpleActor
+import zio.*
+import zio.NonEmptyChunk
+
+case class EditorTechniqueError(
+    id:           BundleName,
+    version:      Version,
+    name:         String,
+    errorMessage: String
+)
+
+sealed trait CompilationStatus
+case object CompilationStatusAllSuccess                                                    extends CompilationStatus
+case class CompilationStatusErrors(techniquesInError: NonEmptyChunk[EditorTechniqueError]) extends CompilationStatus
+
+/**
+  * Update and get latest global techniques compilation status
+  */
+trait ReadTechniqueCompilationStatusService {
+
+  def get(): IOResult[CompilationStatus]
+
+}
+
+/**
+  * Update technique compilation status from technique compilation output and technique info
+  */
+trait WriteTechniqueCompilationStatusService {
+
+  def update(technique: EditorTechnique, output: TechniqueCompilationOutput): UIO[CompilationStatus]
+
+}
+
+/**
+  * Service to update global technique compilation status and get the latest one from filesystem
+  */
+class TechniqueCompilationStatusService(
+    readTechniques:           EditorTechniqueReader,
+    techniqueCompiler:        TechniqueCompiler,
+    compilationStatusService: WriteTechniqueCompilationStatusService
+) extends ReadTechniqueCompilationStatusService {
+
+  override def get(): IOResult[CompilationStatus] = {
```

Comment:
I'm not at ease with that design decision. Here, you have something that is neither functional (ie a kind of pure data pipe function, even if it can fail) or purely stateful (ie with a side effect). 
It means that future usage won't ever know if they need to find of the use place for that service and pipe its result, add a service in the constructor, can use it without changing the world, etc. 

I now get the meaning of that service is "look-up for all compilation output, apply some logic to find the error ones, and notify the interested parties". Not, as I though earlier, "an aggregation of things to build a more complicated logic, perhaps getting data from different points". 
It's almost the second (the implementation is a shiny `for` in disguise), but not exactly, because each time it's called, it has a side effect. So it's not a "read", it's a repository that holds a state - indirectly through the write service, but the write service is mandatory and very specific. 

So, two ways to get out of that: 

- option 1: inverse the read/write relationship (ie write has in its constructor , and make the read just a "gather data / transform data / produce new data" - each time the method is called. 
- option 2: add callback to read, get return `IOResult[Unit]`, write is just one callback

I think the first is better in that case. The second is better when you have a real repository.